File: py/engit/_pull.py
"""engit pull command — git pull on one or more envoy bundle checkouts.

Resolves bundle IDs via ``ENVOY_BNDL_ROOTS`` and runs ``git pull`` in each
repo using Python's :mod:`subprocess` with ``cwd`` set to the bundle path —
no directory changes to the calling process.

Usage::

    engit pull gt:python
    engit pull gt:python gt:globals
    engit pull *
    engit pull * --rebase --dry-run

"""

from __future__ import annotations

import sys
from pathlib import Path

from ._exceptions import EngitError, GitError
from ._git import isGitRepo
from ._git import pull as git_pull


def _resolveSpecs(specs: list[str]) -> list[tuple[str, Path]]:
    """Resolve bundle specs to ``(bndlid, path)`` pairs.

    Args:
        specs: List of bundle ID strings (e.g. ``['gt:python', 'gt:globals']``)
            or ``['*']`` to discover all bundles from ``ENVOY_BNDL_ROOTS``.

    Returns:
        Ordered list of ``(bndlid, path)`` 2-tuples, one per bundle.

    Raises:
        EngitError: If ``*`` is used and ``ENVOY_BNDL_ROOTS`` is not set, or if
            a named bundle ID cannot be resolved.

    """
    try:
        from envoy._discovery import Bundle, discoverBundlesAuto
        from envoy._exceptions import WrapperError
    except ImportError as exc:
        raise EngitError(
            f"Cannot import envoy bundle discovery: {exc}\nEnsure the envoy package is on sys.path."
        ) from exc

    if specs == ['*']:
        bundles = discoverBundlesAuto()
        if not bundles:
            raise EngitError(
                "No bundles discovered. Is ENVOY_BNDL_ROOTS set and pointing to bundle checkouts?"
            )
        return [(b.bndlid, b.root) for b in bundles]

    pairs: list[tuple[str, Path]] = []
    for spec in specs:
        try:
            bundle = Bundle(spec)
        except WrapperError as exc:
            raise EngitError(str(exc)) from exc
        except ValueError as exc:
            raise EngitError(str(exc)) from exc
        pairs.append((bundle.bndlid, bundle.path))
    return pairs


def runPull(
    specs: list[str],
    *,
    remote: str = 'origin',
    rebase: bool = False,
    dry_run: bool = False,
) -> None:
    """Pull one or more envoy bundle checkouts.

    Resolves each spec to a bundle path and runs ``git pull`` in that
    directory via subprocess (``cwd=<bundle_path>``).  The calling process's
    working directory is never changed.

    Args:
        specs: Bundle IDs (e.g. ``['gt:python']``) or ``['*']`` for all
            bundles discovered from ``ENVOY_BNDL_ROOTS``.
        remote: Remote to pull from. Defaults to ``'origin'``.
        rebase: Pass ``--rebase`` to ``git pull``.
        dry_run: Print what would be pulled without running git.

    Raises:
        EngitError: If bundle resolution fails (e.g. ``ENVOY_BNDL_ROOTS`` not
            set, unknown bundle ID).  In multi-bundle mode per-bundle git
            failures are collected and reported as a summary rather than
            raising immediately.

    """
    bundles = _resolveSpecs(specs)
    multi = len(bundles) > 1

    if dry_run:
        action = f'git pull{" --rebase" if rebase else ""} {remote}'
        if multi:
            print(f'Would pull {len(bundles)} bundle(s) [{action}]:')
            for bndlid, path in bundles:
                print(f'  {bndlid:<20}  {path}')
        else:
            bndlid, path = bundles[0]
            print(f'Would pull {bndlid}  ({path})\n  Command: {action}')
        return

    if multi:
        print(f'Pulling {len(bundles)} bundle(s)...')

    failures: list[tuple[str, str]] = []
    col = max(len(b) for b, _ in bundles) + 2

    for bndlid, path in bundles:
        if not isGitRepo(path):
            msg = 'skipped (not a git repo)'
            if multi:
                print(f'  {bndlid:<{col}} ⚠  {msg}')
            else:
                print(f'{bndlid}: {msg}', file=sys.stderr)
            continue

        if multi:
            print(f'  {bndlid:<{col}}', end='', flush=True)
        else:
            print(f'Pulling {bndlid}  ({path})')

        try:
            output = git_pull(remote=remote, rebase=rebase, cwd=path)
            first_line = output.splitlines()[0] if output.strip() else 'Done.'
            if multi:
                print(f' ✓  {first_line}')
            else:
                print(output if output.strip() else 'Already up to date.')
        except GitError as exc:
            err = str(exc).splitlines()[0]
            if multi:
                print(f' ✗  {err}')
                failures.append((bndlid, str(exc)))
            else:
                raise EngitError(str(exc)) from exc

    if multi:
        succeeded = len(bundles) - len(failures)
        print(f'\nDone. {succeeded} succeeded, {len(failures)} failed.')
        if failures:
            print('  FAILED:')
            for bndlid, err in failures:
                print(f'    {bndlid}: {err.splitlines()[0]}')

File: rust/engit-core/src/pull.rs
//! Pull one or more envoy bundle checkouts.

use std::path::{Path, PathBuf};

use envoy_core::discovery::{discover_bundles_auto, Bundle};

use crate::error::{EngitError, Result};
use crate::git::{is_git_repo, pull as git_pull};

pub(crate) fn resolve_specs(specs: &[String]) -> Result<Vec<(String, PathBuf)>> {
    if specs.len() == 1 && specs[0] == "*" {
        let bundles = discover_bundles_auto()?;
        if bundles.is_empty() {
            return Err(EngitError::Engit(String::from(
                "No bundles discovered. Is ENVOY_BNDL_ROOTS set and pointing \
to bundle checkouts?",
            )));
        }

        return Ok(bundles
            .into_iter()
            .map(|bundle| (bundle.bndlid(), bundle.root))
            .collect());
    }

    let mut pairs = Vec::new();
    for spec in specs {
        let bundle = Bundle::new(Path::new(spec), None)?;
        pairs.push((bundle.bndlid(), bundle.path().to_path_buf()));
    }

    Ok(pairs)
}

/// Pull one or more envoy bundle checkouts.
pub fn run_pull(specs: &[String], remote: &str, rebase: bool, dry_run: bool) -> Result<()> {
    let bundles = resolve_specs(specs)?;
    let multi = bundles.len() > 1;

    if dry_run {
        let action = format!("git pull{} {remote}", if rebase { " --rebase" } else { "" });
        if multi {
            println!("Would pull {} bundle(s) [{action}]:", bundles.len());
            for (bndlid, path) in &bundles {
                println!("  {bndlid:<20}  {}", path.display());
            }
        } else if let Some((bndlid, path)) = bundles.first() {
            println!("Would pull {bndlid}  ({})", path.display());
            println!("  Command: {action}");
        }
        return Ok(());
    }

    if multi {
        println!("Pulling {} bundle(s)...", bundles.len());
    }

    let mut failures = Vec::new();
    let width = bundles
        .iter()
        .map(|(bndlid, _)| bndlid.len())
        .max()
        .unwrap_or(0)
        + 2;

    for (bndlid, path) in &bundles {
        if !is_git_repo(Some(path)) {
            let message = "skipped (not a git repo)";
            if multi {
                println!("  {bndlid:<width$} ⚠  {message}");
            } else {
                eprintln!("{bndlid}: {message}");
            }
            continue;
        }

        if multi {
            print!("  {bndlid:<width$}");
        } else {
            println!("Pulling {bndlid}  ({})", path.display());
        }

        match git_pull(remote, rebase, Some(path)) {
            Ok(output) => {
                let first_line = output
                    .lines()
                    .next()
                    .filter(|line| !line.trim().is_empty())
                    .unwrap_or("Done.");
                if multi {
                    println!(" ✓  {first_line}");
                } else if output.trim().is_empty() {
                    println!("Already up to date.");
                } else {
                    println!("{output}");
                }
            }
            Err(error) => {
                let first_line = error
                    .to_string()
                    .lines()
                    .next()
                    .unwrap_or_default()
                    .to_string();
                if multi {
                    println!(" ✗  {first_line}");
                    failures.push((bndlid.clone(), error.to_string()));
                } else {
                    return Err(EngitError::Engit(error.to_string()));
                }
            }
        }
    }

    if multi {
        let succeeded = bundles.len().saturating_sub(failures.len());
        println!("\nDone. {succeeded} succeeded, {} failed.", failures.len());
        if !failures.is_empty() {
            println!("  FAILED:");
            for (bndlid, error) in failures {
                let first_line = error.lines().next().unwrap_or_default();
                println!("    {bndlid}: {first_line}");
            }
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use std::ffi::{OsStr, OsString};
    use std::fs;

    use tempfile::tempdir;

    use super::resolve_specs;
    use crate::BUNDLE_ROOTS_ENV_MUTEX;

    struct EnvVarGuard {
        previous: Option<OsString>,
    }

    impl EnvVarGuard {
        fn set(value: Option<&OsStr>) -> Self {
            let previous = std::env::var_os("ENVOY_BNDL_ROOTS");
            match value {
                Some(value) => std::env::set_var("ENVOY_BNDL_ROOTS", value),
                None => std::env::remove_var("ENVOY_BNDL_ROOTS"),
            }
            Self { previous }
        }
    }

    impl Drop for EnvVarGuard {
        fn drop(&mut self) {
            match &self.previous {
                Some(value) => std::env::set_var("ENVOY_BNDL_ROOTS", value),
                None => std::env::remove_var("ENVOY_BNDL_ROOTS"),
            }
        }
    }

    #[test]
    fn resolves_explicit_and_wildcard_bundle_specs() {
        let _lock = BUNDLE_ROOTS_ENV_MUTEX
            .lock()
            .expect("bundle roots env mutex poisoned");
        let temp = tempdir().expect("failed to create temp dir");
        let bundle_root = temp.path().join("bundles");
        let bundle = bundle_root.join("gt").join("pythoncore");
        fs::create_dir_all(bundle.join(".envoy")).expect("failed to create bundle .envoy");
        fs::create_dir_all(bundle.join(".git")).expect("failed to create bundle .git");

        let joined = std::env::join_paths([bundle_root.as_path()]).expect("failed to join paths");
        let _env_guard = EnvVarGuard::set(Some(joined.as_os_str()));

        let explicit = resolve_specs(&[String::from("gt:pythoncore")])
            .expect("explicit bundle should resolve");
        let wildcard = resolve_specs(&[String::from("*")]).expect("wildcard should resolve");

        assert_eq!(explicit[0].0, "gt:pythoncore");
        assert_eq!(
            explicit[0]
                .1
                .canonicalize()
                .expect("explicit bundle path should canonicalize"),
            bundle
                .canonicalize()
                .expect("expected bundle path should canonicalize")
        );
        assert_eq!(wildcard[0].0, "gt:pythoncore");
    }
}
